### ISSUE TITLE
Conditional rule with conda environment

### ISSUE THREAD

**Issue by mevers** [snakemake/snakemake#163]
Hi.

Following advice given by Johannes Köster in [Add complex conditional file dependency
](https://bitbucket.org/snakemake/snakemake/issues/37/add-complex-conditional-file-dependency) on how to include a conditional I've got the following conditional rule

```python
rule bowtie2_align_reads:
    input:
        idx = IDX,
        reads = lambda wildcards: expand(join(
            config["fastqdir"],
            "{file}"),
            file = config["units"][wildcards.sample])
    output:
        join(
            config["analysisdir"],
            "{reference_version}/alignment/{sample}.bam")
    log:
        "logs/bowtie2_{reference_version}_{sample}.log"
    params:
        cmd      = config["bowtie2"]["cmd"],
        ref      = re.sub(".fa", "", REF),
        in_fmt   = config["bowtie2"]["in_fmt"],
        phred    = config["bowtie2"]["phred"],
        maxins   = config["bowtie2"]["maxins"],
        mismatch = config["bowtie2"]["mismatch"],
        threads  = config["bowtie2"]["threads"]
    version: "1.0"
    run:
        if (len(input.reads) == 2):
            shell(" \
                {params.cmd} \
                {params.phred} \
                --no-mixed \
                --no-discordant \
                --maxins {params.maxins} \
                -N {params.mismatch} \
                --threads {params.threads} \
                -x {params.ref} \
                -1 {input.reads[0]} \
                -2 {input.reads[1]} \
                2> {log} \
                | samtools view -bS -F4 - > {output} \
            ")
        elif (len(input.reads) == 1):
            shell(" \
                {params.cmd} \
                {params.phred} \
                -N {params.mismatch} \
                --threads {params.threads} \
                -x {params.ref} \
                -U {input.reads} \
                2> {log} \
                | samtools view -bS -F4 - > {output} \
            ")
```

In an attempt to "robustify" the workflow I am trying to use conda environments per rule. However, when I add 

```python
conda:
    "envs/bowtie2.yaml"
``` 

I get the following error
```
RuleException in line 42 of /home/mevers/Projects/ChIPseq_UBF_reanalysis/snakemake/rules/alignment.smk:
Conda environments are only allowed with shell, script, or wrapper directives (not with run).
```

My questions are:

1. Is this a bug or intended behaviour? The conditional rule uses two `shell` commands within an `if-elif` construct nested in a `run` directive. So while the top-level directive is `run`, the actual rule decides on one of two `shell` calls.

2.  If this is intended behaviour, what would be the canonical way to include a conditional rule while making use of a conda environment?

Thanks,
Maurits

**Comment by SichongP** [comment]
How about putting your commands in a python script or a bash script? You can invoke python script with `script` directive or bash script with `shell` directive like this: `shell: 'bash myScript.sh --idx {input.idx} --output {output} --log {log} ...'`

**Comment by mevers** [comment]
@SichongP Yes that would be a possibility but at least as far as the Python script is concerned is not very neat.

Having to refer to a separate Python script just to spawn a `bowtie2` sub-process with e.g. `subprocess.run` or `subprocess.Popen` seems somewhat convoluted (there's a related [post on SeqAnswers](http://seqanswers.com/forums/showthread.php?t=72515) where this is being discouraged).

That's why I'm interested in knowing whether there's a more canonical within-`snakemake` kind of solution.

**Comment by SichongP** [comment]
@mevers I would argue that this is a different use case from the post you referenced. It was talking about using snakemake instead of a Python script to manage a workflow with many iterations of same process, hence "not reinventing the wheels". In your use case you're trying to invoke different commands based on input in each iteration of the same process. Seems to me that a script is a fitting solution.

However, if you really don't want to use a script, maybe you can try with a shell block, where you can pass input to a bash array, checking the length of your array, and then invoke different commands accordingly?
I've never used an array in a shell block in snakemake so it may need some experimenting but theoretically it should work.

**Comment by mevers** [comment]
@SichongP I agree with you in that the SeqAnswers post is not directly related. It does however raise the point (perhaps somewhat indirectly) that spawning subprocesses in Python should generally be avoided.

Perhaps we're missing each others points. Let me explain/rephrase: I find it unsatisfactory that the "conditional `run`" solution works fine *without* `conda` environment but fails *with* `conda` environment. After all, the "conditional `run`" solution seems to be a/the recommended way to implement conditional input file dependencies.

A `bash` script may be the solution; it would require re-writing parts of my workflow to encapsulate all conditional file dependencies in separate `bash` scripts, just to allow for local `conda` environments.

**Comment by SichongP** [comment]
@mevers So by shell block, I was talking about something like this:
```
rule bowtie2_align_reads:
    input:
        idx = IDX,
        reads = lambda wildcards: expand(join(
            config["fastqdir"],
            "{file}"),
            file = config["units"][wildcards.sample])
...
...
    shell:
     """
     input_reads=({input.reads})
     if [ "${{#input_reads[@]}}" -eq 2 ]; then
         {params.cmd} {params.phred} --no-mixed ... ## bowtie command for 2 input files
     elif [ "${{#input_reads[@]}}" -eq 1 ]; then
         ## bowtie command for 1 input file
     else
         ## error handling
     fi
     """
```
I just tested with a minimal workflow and it should work as intended. `shell` directive does work with conda so this should be the solution to your problem I think.
As to why `run` directive is not compatible with conda env, I'm not exactly sure. But this does get mentioned in the documentation specifically and in source code it is specifically checked so my guess is that it has something to do with the way `run` or conda env is implemented in snakemake.

**Comment by tbooth** [comment]
I've removed the 'bug' tag since my understanding, as per @SichongP, is that this behaviour is intentional and that moving the if/else into the shell block is the current preferred way to implement these cases. However there may still be an enhancement and/or extra documentation to consider.

**Comment by mevers** [comment]
Sorry for the delay in my response @SichongP ; we've had a few intense weeks down here in Australia. 

Indeed, your `shell` solution is what I was after. Thanks for your help.

**Comment by marykthompson** [comment]
In case anyone finds this googling like I did, I just want to point out that the shell block solution where you test the length of the input might not work depending on what you want to run afterwards in your shell block. In my case, I need use {input.reads[0]} for single-end samples and both {input.reads[0]} and {input.reads[1]} for the paired-end samples. This means that when running a single-end sample, Snakemake will still try to substitute in a value for {input.reads[1]} into the shell command even though it doesn't exist. Then it throws a list index out of range error and doesn't run.

**Comment by Hocnonsense** [comment]
This should be fixed via #2289
